src/ucp/wireup/wireup.c

```diff
@@ -210,9 +215,178 @@ static void ucp_wireup_msg_get_peer_name(const ucp_wireup_msg_t *msg,
     peer_name[length] = '\0';
 }
 
-static struct sockaddr * ucp_wireup_msg_get_aux_addr(const ucp_wireup_msg_t *msg)
+static void ucp_wireup_msg_get_tl_name(const ucp_wireup_msg_t *msg,
+                                       char *tl_name, size_t max)
 {
-    return (struct sockaddr *)((void*)(msg + 1) + msg->addr_len);
+    size_t length = ucs_min(max - 1, msg->tl_name_len);
+    memcpy(tl_name, (const char*)(msg + 1) + msg->peer_name_len, length);
+    tl_name[length] = '\0';
+}
+
+static struct sockaddr * ucp_wireup_msg_get_aux_addr(ucp_wireup_msg_t *msg)
+{
+    void *msg_data = msg + 1;
+    return msg_data + msg->peer_name_len + msg->tl_name_len + msg->addr_len;
+}
+
+static double ucp_aux_score_func(ucp_worker_h worker, uct_iface_attr_t *iface_attr)
+{
+    if (!(iface_attr->cap.flags & UCT_IFACE_FLAG_AM_BCOPY) || /* Need to use it for wireup messages */
+        !(iface_attr->cap.flags & UCT_IFACE_FLAG_CONNECT_TO_IFACE) || /* Should connect immediately */
+        !(iface_attr->cap.flags & UCT_IFACE_FLAG_AM_CB_ASYNC) || /* Should progress asynchronously */
+        !(iface_attr->cap.flags & UCT_IFACE_FLAG_PENDING) /* Should support pending operations */ )
+    {
+        return 0.0;
+    }
+
+    return (1e-3 / iface_attr->latency) +
+           (1e3 * ucs_max(iface_attr->cap.am.max_bcopy, iface_attr->cap.am.max_short));
+}
+
+static double ucp_runtime_score_func(ucp_worker_h worker, uct_iface_attr_t *iface_attr)
+{
+    ucp_context_t *context = worker->context;
+    uint64_t flags;
+
+    flags = 0;
+
+    if (iface_attr->cap.flags & UCT_IFACE_FLAG_CONNECT_TO_EP) {
+        flags |= UCT_IFACE_FLAG_AM_BCOPY;
+    }
+
+    if (context->config.features & UCP_FEATURE_TAG) {
+        flags |= UCT_IFACE_FLAG_AM_SHORT | UCT_IFACE_FLAG_PENDING | UCT_IFACE_FLAG_AM_CB_SYNC;
+    }
+
+    if (context->config.features & UCP_FEATURE_RMA) {
+        /* TODO remove this requirement once we have RMA emulation */
+        flags |= UCT_IFACE_FLAG_PUT_SHORT | UCT_IFACE_FLAG_PUT_BCOPY |
+                 UCT_IFACE_FLAG_GET_BCOPY;
+    }
+
+    if (context->config.features & UCP_FEATURE_AMO32) {
+        /* TODO remove this requirement once we have SW atomics */
+        flags |= UCT_IFACE_FLAG_ATOMIC_ADD32 | UCT_IFACE_FLAG_ATOMIC_FADD32 |
+                 UCT_IFACE_FLAG_ATOMIC_SWAP32 | UCT_IFACE_FLAG_ATOMIC_CSWAP32;
+    }
+
+    if (context->config.features & UCP_FEATURE_AMO64) {
+        /* TODO remove this requirement once we have SW atomics */
+        flags |= UCT_IFACE_FLAG_ATOMIC_ADD64 | UCT_IFACE_FLAG_ATOMIC_FADD64 |
+                 UCT_IFACE_FLAG_ATOMIC_SWAP64 | UCT_IFACE_FLAG_ATOMIC_CSWAP64;
+    }
```

Comment:
@MattBBaker I think Aaron have been complaining that this logic does not pick up smsg+fma ?